src/rust/cryptography-x509-validation/src/certificate.rs

```diff
@@ -0,0 +1,80 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+use asn1::ObjectIdentifier;
+use cryptography_x509::{
+    certificate::Certificate,
+    extensions::{Extension, Extensions},
+};
+
+use crate::{backend::CryptoOps, types::Name};
+
+#[derive(Debug, PartialEq)]
+pub enum CertificateError {
+    DuplicateExtension(asn1::ObjectIdentifier),
+    Malformed(asn1::ParseError),
+}
+
+impl From<asn1::ParseError> for CertificateError {
+    fn from(value: asn1::ParseError) -> Self {
+        CertificateError::Malformed(value)
+    }
+}
+
+pub(crate) trait CertificateExt: private::Sealed {
+    fn is_self_issued(&self) -> bool;
+    fn is_self_signed<B: CryptoOps>(&self, ops: &B) -> bool;
```

Comment:
@alex How do you want to handle this extensional trait? It relies on `CryptoOpts` here, so removing it by merging its APIs into the main `impl Certificate` probably means moving `CryptoOps` into `cryptography_x509` rather than its current `cryptography_x509_validation`.

Reply:
Can we just have a non-method function for these?